### PR TITLE
UI: Use QScopedPointers for dialogs

### PR DESCRIPTION
**Pull request by cg2121**
### Description
With QScopedPointers, the dialogs are automatically deleted
when going out of scope.

### Motivation and Context
Automating deletion is nice.

### How Has This Been Tested
Open and closed all of the dialogs and closed OBS to make sure nothing broke.

### Types of changes
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by cg2121** [comment]
I've updated this to use QScopedPointers instead.

**Comment by jp9000** [comment]
I want to say that this is fundamentally a positive change that I would normally be happy to merge. Things should always be designed around scope and RAII in C++, which is the intended way to use C++. The intent is on point and correct, and there's nothing really to criticize in this submission outside of a misunderstanding of why the code was the way it was.

There are some caveats to particular situation which makes this change not quite as trivial to accomplish, and I'll explain what I mean.

Firstly, there is the issue of the removal of Qt::WA_DeleteOnClose. The Qt::WA_DeleteOnClose ensures that when a dialog is closed, it is fully deleted. When the flag is not present, the dialog still exists internally, but is instead hidden when closed; this is done as an optimization by Qt because as the hotkey section of the settings window shows us, creation of widgets does not have very good performance in Qt, and it creates a large number of allocations and objects. Removing this flag means that dialogs that are holding source references for whatever reason (even if temporarily) will continue to hold those references until the program closes, effectively making those sources perpetually exist until the end of the program whether anyone likes it or not. So that flag exists specifically for this situation, to ensure data within the object is cleaned up when the dialog is closed (particularly OBS object references). Additionally, this makes using QScopedPointer dangerous, as the object can be deleted before it's actually deleted. In retrospect, QPointer really is actually the correct choice for these particular objects because QPointer basically holds a weak reference to the Qt object instead of a strong reference, thus it knows when the object has been deleted elsewhere. (I just wish they hadn't named it QPointer)

Secondly, there is the unfortunate issue of order of operations in the destructor. You'll notice in the ~OBSBasic() destructor that we delete the objects manually, which you removed. This is not done because we actually need to delete the objects; technically we do not have to delete them because they are already children of the main window, and will be automatically destroyed with all other children once the ~QWidget() destructor executes. Instead, this is done because we want to ensure they're deleted before they reach the QApplication::sendPostedEvents() call that you'll notice is below it (with a comment stating why it's there). This is a completely understandable misconception to make and there should have been a comment in the code to indicate that those deletes are being done for order of operation reasons rather than simply to clean up some objects. In hindsight, there mat likely be a better and more clean way to accomplish the same thing. Or at least a comment.

Basically, although this PR (sans the unnecessary `->data()` calls) is how you *should* do it under normal circumstances. However, the complication is primarily OBS object reference holders and ensuring they're fully released before the actual shutdown.

I'm closing this primarily for these reasons. We want to control the timing of the destruction for these objects on purpose. Both when the dialogs are closed by the user, and at a specific point in the destructor. Your change, although a fundamentally good change, unfortunately falls pray to that completely understandable misconception. I can really only blame myself for not trying to find a cleaner way at the time.
